java/org/apache/jasper/compiler/Generator.java

```diff
@@ -81,6 +83,13 @@
 
     private static final Class<?>[] OBJECT_CLASS = { Object.class };
 
+    //context param to enable or disable the excess white space trimming.
+    private static final String JSP_WHITE_SPACE_TRIMMING = "JSPWhiteSpaceTrimming";
+
+    private static final Pattern PRE_TAG_PATTERN = Pattern.compile("(?s).*(<pre>|</pre>).*");
+
+    private static final Pattern BLANK_LINE_PATTERN = Pattern.compile("(\\s*(\\n|\\r)+\\s*)");
```

Comment:
We should provide a set of test cases that exercises these matchers.

Reply:
I think we need input here, if we should add another webapp - for testing this change.

Reply:
Can we not add tests here: https://github.com/apache/tomcat/blob/master/test/org/apache/jasper/compiler/TestGenerator.java?

Reply:
I think we might need to add another JSP file, with the patterns. I will look into it.Thanks